simulator/core/perf_sim.cpp

```diff
@@ -145,63 +145,96 @@ void PerfMIPS::clock_decode( int cycle)
 
     /* receive flush signal */
     const bool is_flush = rp_decode_flush->is_ready( cycle) && rp_decode_flush->read( cycle);
-
+    IfIdData _data;
     /* branch misprediction */
     if ( is_flush)
     {
         /* ignoring the upcoming instruction as it is invalid */
-        decode_data = rp_fetch_2_decode->read( cycle);
+        _data = rp_fetch_2_decode->read( cycle);
+	    if(rp_decode_2_decode->is_ready(cycle))
+        {
+            rp_decode_2_decode->read(cycle);
+        }
 
-        is_anything_to_decode = false;
         sout << "flush\n";
         return;
     }
-
-    if ( !is_anything_to_decode)
+    /* check if there is something to process */
+    if( !rp_fetch_2_decode->is_ready( cycle) && !rp_decode_2_decode->is_ready( cycle))
     {
-        /* acquiring data from fetch */
-        is_anything_to_decode = rp_fetch_2_decode->is_ready( cycle);
-        if ( is_anything_to_decode)
+        sout << "bubble\n";
+        return;
+    }
+    if(rp_decode_2_decode->is_ready( cycle))
+    {
+    	auto instr = rp_decode_2_decode->read( cycle);
+	    
+        rp_fetch_2_decode->ignore( cycle);
+	    
+        if ( rf->check_sources( instr))
         {
-            decode_data = rp_fetch_2_decode->read( cycle);
+            rf->read_sources( &instr);
+
+            wp_decode_2_execute->write( instr, cycle);
+
+        /* log */
+        sout << instr << std::endl;
         }
+   	    else // data hazard, stalling pipeline
+   	    {
+        wp_decode_2_fetch_stall->write( true, cycle);
+	    wp_decode_2_decode->write( instr, cycle);
+        sout << instr << " (data hazard)\n";
+   	    }
+
     }
     else
     {
-        /* ignore data from port -- to suppress loss messages */
-        rp_fetch_2_decode->ignore( cycle);
-    }
+    if(rp_fetch_2_decode->is_ready( cycle)) {
+        _data = rp_fetch_2_decode->read( cycle);
+        FuncInstr instr( _data.raw, 
+                _data.PC, 
+                _data.predicted_taken, 
+                _data.predicted_target);
+         if ( rf->check_sources( instr))
+         {
+            rf->read_sources( &instr);
+        
+            wp_decode_2_execute->write( instr, cycle);
+
+            /* log */
+            sout << instr << std::endl;
+         }
+         else // data hazard, stalling pipeline
+         {
+            wp_decode_2_fetch_stall->write( true, cycle);
+            wp_decode_2_decode->write(instr, cycle);
+            sout << instr << " (data hazard)\n";
+         }
+    
 
-    /* check if there is something to process */
-    if ( !is_anything_to_decode)
-    {
-        sout << "bubble\n";
-        return;
     }
+    }
+}
 
-    FuncInstr instr( decode_data.raw,
-                     decode_data.PC,
-                     decode_data.predicted_taken,
-                     decode_data.predicted_target);
 
     /* TODO: replace all this code by introducing Forwarding unit */
-    if ( rf->check_sources( instr))
+    /*if ( rf->check_sources( instr))
```

Comment:
Please do not leave commented code as it pollutes search results. All the changes are stored in Git.